README.md

[comment]: # (Start Badges)

[![Build Status](https://travis-ci.org/47degrees/github4s.svg?branch=master)](https://travis-ci.org/47degrees/github4s) [![Join the chat at https://gitter.im/47degrees/github4s](https://badges.gitter.im/47degrees/github4s.svg)](https://gitter.im/47degrees/github4s?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) [![codecov.io](http://codecov.io/gh/47degrees/github4s/branch/master/graph/badge.svg)](http://codecov.io/gh/47degrees/github4s) [![Maven Central](https://img.shields.io/badge/maven%20central-0.23.0-green.svg)](https://oss.sonatype.org/#nexus-search;gav~com.47deg~github4s*) [![License](https://img.shields.io/badge/license-Apache%202-blue.svg)](https://raw.githubusercontent.com/47degrees/github4s/master/LICENSE) [![Latest version](https://img.shields.io/badge/github4s-0.23.0-green.svg)](https://index.scala-lang.org/47degrees/github4s) [![GitHub Issues](https://img.shields.io/github/issues/47degrees/github4s.svg)](https://github.com/47degrees/github4s/issues)

[comment]: # (End Badges)

Github4s
=============

**Github4s** is a GitHub API wrapper written in Scala.

## Installation

To get started with SBT, simply add the following to your build.sbt file.

[comment]: # (Start Replace)

```scala
libraryDependencies += "com.47deg" %% "github4s" % "0.23.0"
```

[comment]: # (End Replace)

## Github4s in the wild

If you wish to add your library here please consider a PR to include it in the list below.

★ | ★ | ★
--- | --- | ---
![scala-exercises](https://www.scala-exercises.org/assets/images/navbar_brand.svg) | [**scala-exercises**](https://www.scala-exercises.org/) | Scala Exercises is an Open Source project for learning different technologies based in the Scala Programming Language.
| | [**dashing**](https://github.com/benfradet/dashing) | Dashing is a collection of dashboards to monitor the health of an open source organization.
| | [**cla-bot**](https://github.com/snowplow-incubator/cla-bot) | Bot making sure external contributors sign a CLA.

[comment]: # (Start Copyright)
# Copyright

Github4s is designed and developed by 47 Degrees

Copyright (C) 2016-2020 47 Degrees. <http://47deg.com>

[comment]: # (End Copyright)

CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/47degrees/github4s/tree/HEAD)

[Full Changelog](https://github.com/47degrees/github4s/compare/v0.23.0...HEAD)

🚀 **Features**

- Better exception messages/classes [\#118](https://github.com/47degrees/github4s/issues/118)

**Closed issues:**

- Integration tests are run when the token is not present [\#396](https://github.com/47degrees/github4s/issues/396)
- Support for private GitHub Enterprise installations [\#379](https://github.com/47degrees/github4s/issues/379)
- Shutting down connection pool info messages [\#371](https://github.com/47degrees/github4s/issues/371)
- Clarify error handling design [\#363](https://github.com/47degrees/github4s/issues/363)
- Remove list statuses test flakiness [\#300](https://github.com/47degrees/github4s/issues/300)
- Traversing with Pagination [\#285](https://github.com/47degrees/github4s/issues/285)
- Wrong API endpoint for listUserRepos [\#255](https://github.com/47degrees/github4s/issues/255)
- Add create milestone [\#248](https://github.com/47degrees/github4s/issues/248)

**Merged pull requests:**

- Renames GH token env var [\#405](https://github.com/47degrees/github4s/pull/405) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Updated and delete milestone [\#402](https://github.com/47degrees/github4s/pull/402) ([anamariamv](https://github.com/anamariamv))
- Don't propagate dummy configuration to integration tests [\#401](https://github.com/47degrees/github4s/pull/401) ([BenFradet](https://github.com/BenFradet))
- Get single milestone [\#400](https://github.com/47degrees/github4s/pull/400) ([anamariamv](https://github.com/anamariamv))
- Fix API documentation link for the user repos endpoint [\#399](https://github.com/47degrees/github4s/pull/399) ([BenFradet](https://github.com/BenFradet))
- Renames Token Env Var [\#397](https://github.com/47degrees/github4s/pull/397) ([juanpedromoreno](https://github.com/juanpedromoreno))
- API: Create milestone [\#391](https://github.com/47degrees/github4s/pull/391) ([anamariamv](https://github.com/anamariamv))
- feat: add Create / Update / Delete requests [\#390](https://github.com/47degrees/github4s/pull/390) ([kalexmills](https://github.com/kalexmills))
- Update http4s-blaze-client, http4s-circe to 0.21.3 [\#387](https://github.com/47degrees/github4s/pull/387) ([scala-steward](https://github.com/scala-steward))
- Update mockserver-netty to 5.10 [\#386](https://github.com/47degrees/github4s/pull/386) ([scala-steward](https://github.com/scala-steward))
- Introduces custom GitHub configuration [\#384](https://github.com/47degrees/github4s/pull/384) ([satorg](https://github.com/satorg))
- Update sbt-microsites to 1.1.5 [\#383](https://github.com/47degrees/github4s/pull/383) ([scala-steward](https://github.com/scala-steward))
- Update sbt-mdoc to 2.1.5 [\#382](https://github.com/47degrees/github4s/pull/382) ([scala-steward](https://github.com/scala-steward))
- Update sbt-microsites to 1.1.4 [\#380](https://github.com/47degrees/github4s/pull/380) ([scala-steward](https://github.com/scala-steward))
- Mergify: configuration update [\#378](https://github.com/47degrees/github4s/pull/378) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Tag test for listing starrers as integration [\#376](https://github.com/47degrees/github4s/pull/376) ([BenFradet](https://github.com/BenFradet))
- Update sbt-mdoc to 2.1.4 [\#375](https://github.com/47degrees/github4s/pull/375) ([scala-steward](https://github.com/scala-steward))
- Update sbt-org-policies to 0.13.2 [\#374](https://github.com/47degrees/github4s/pull/374) ([scala-steward](https://github.com/scala-steward))
- Require an http4s client to provide more flexibility [\#373](https://github.com/47degrees/github4s/pull/373) ([BenFradet](https://github.com/BenFradet))

## [v0.23.0](https://github.com/47degrees/github4s/tree/v0.23.0) (2020-03-16)

[Full Changelog](https://github.com/47degrees/github4s/compare/v0.22.0...v0.23.0)

🚀 **Features**

- Integrate a new API: Get the list of cards for a project \(given a column id\) [\#361](https://github.com/47degrees/github4s/issues/361)
- Integrate a new API: Get the list of projects in a repository [\#360](https://github.com/47degrees/github4s/issues/360)
- Integrate a new API: Get the list of milestones in a project [\#340](https://github.com/47degrees/github4s/issues/340)
- Integrate a new API: Get the list of columns in a project [\#339](https://github.com/47degrees/github4s/issues/339)
- Integrate a new API: Get the list of projects in an organization [\#338](https://github.com/47degrees/github4s/issues/338)

**Closed issues:**

- Remove unnecessary folder structure [\#357](https://github.com/47degrees/github4s/issues/357)
- Include an example using F in the docs [\#356](https://github.com/47degrees/github4s/issues/356)
- Integrate a new API: Get the list of labels in a project [\#341](https://github.com/47degrees/github4s/issues/341)

**Merged pull requests:**

- Releases 0.23.0 [\#372](https://github.com/47degrees/github4s/pull/372) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Migrates from 47deg to 47degrees GH organization [\#370](https://github.com/47degrees/github4s/pull/370) ([juanpedromoreno](https://github.com/juanpedromoreno))
- Update sbt-mdoc to 2.1.3 [\#369](https://github.com/47degrees/github4s/pull/369) ([scala-steward](https://github.com/scala-steward))
- Introduce sbt-tpolecat [\#368](https://github.com/47degrees/github4s/pull/368) ([BenFradet](https://github.com/BenFradet))
- Rework GHResponse to incorporate useful data on unhappy path [\#367](https://github.com/47degrees/github4s/pull/367) ([BenFradet](https://github.com/BenFradet))
- Update sbt-microsites to 1.1.3 [\#366](https://github.com/47degrees/github4s/pull/366) ([scala-steward](https://github.com/scala-steward))
- List repository projects  [\#365](https://github.com/47degrees/github4s/pull/365) ([anamariamv](https://github.com/anamariamv))

project/ProjectPlugin.scala
import com.typesafe.sbt.site.SitePlugin.autoImport._
import microsites._
import microsites.MicrositesPlugin.autoImport._
import sbt.Keys._
import sbt._
import sbtorgpolicies.OrgPoliciesKeys.orgBadgeListSetting
import sbtorgpolicies.OrgPoliciesPlugin
import sbtorgpolicies.OrgPoliciesPlugin.autoImport._
import sbtorgpolicies.templates.badges._
import sbtorgpolicies.runnable.syntax._
import scoverage.ScoverageKeys
import scoverage.ScoverageKeys._
import mdoc.MdocPlugin.autoImport._
import sbtorgpolicies.model.GitHubSettings

object ProjectPlugin extends AutoPlugin {

  override def trigger: PluginTrigger = allRequirements

  override def requires: Plugins = OrgPoliciesPlugin

  object autoImport {

    lazy val V = new {
      val base64: String     = "0.2.9"
      val cats: String       = "2.1.1"
      val catsEffect: String = "2.1.1"
      val circe: String      = "0.13.0"
      val paradise: String   = "2.1.1"
      val simulacrum: String = "0.19.0"
      val scala212: String   = "2.12.10"
      val scala213: String   = "2.13.1"
      val http4s: String     = "0.21.3"
      val scalamock: String  = "4.4.0"
      val scalaTest: String  = "3.1.1"
      val silencer: String   = "1.6.0"
    }

    lazy val micrositeSettings = Seq(
      micrositeName := "Github4s",
      micrositeDescription := "Github API wrapper written in Scala",
      micrositeBaseUrl := "github4s",
      micrositeDocumentationUrl := "docs",
      micrositeGithubOwner := "47degrees",
      micrositeGithubRepo := "github4s",
      micrositeAuthor := "Github4s contributors",
      micrositeCompilingDocsTool := WithMdoc,
      micrositePushSiteWith := GitHub4s,
      micrositeOrganizationHomepage := "https://github.com/47degrees/github4s/blob/master/AUTHORS.md",
      micrositePalette := Map(
        "brand-primary"   -> "#3D3832",
        "brand-secondary" -> "#f90",
        "white-color"     -> "#FFFFFF"
      ),
      micrositeExtraMdFiles := Map(
        file("CHANGELOG.md") -> ExtraMdFileConfig(
          "changelog.md",
          "page",
          Map(
            "title"     -> "Changelog",
            "section"   -> "home",
            "position"  -> "3",
            "permalink" -> "changelog"
          )
        )
      ),
      micrositeExtraMdFilesOutput := mdocIn.value,
      includeFilter in makeSite := "*.html" | "*.css" | "*.png" | "*.jpg" | "*.gif" | "*.js" | "*.swf" | "*.md" | "*.svg",
      scalacOptions ~= (_ filterNot Set("-Ywarn-unused-import", "-Xlint", "-Xfatal-warnings").contains)
    )

    lazy val coreDeps = Seq(
      libraryDependencies ++= Seq(
        %%("cats-core", V.cats),
        %%("cats-free", V.cats),
        %%("simulacrum", V.simulacrum),
        %%("circe-core", V.circe),
        %%("circe-generic", V.circe),
        "io.circe" %% "circe-literal" % V.circe,
        %%("base64", V.base64),
        "org.http4s"                 %% "http4s-blaze-client" % V.http4s,
        "org.http4s"                 %% "http4s-circe" % V.http4s,
        %%("circe-parser", V.circe)  % Test,
        %%("scalamock", V.scalamock) % Test,
        %%("scalatest", V.scalaTest) % Test,
        "org.mock-server"            % "mockserver-netty" % "5.10" % Test excludeAll ExclusionRule(
          "com.twitter"
        ),
        compilerPlugin("com.github.ghik" % "silencer-plugin" % V.silencer cross CrossVersion.full),
        "com.github.ghik" % "silencer-lib" % V.silencer % Provided cross CrossVersion.full
      ),
      libraryDependencies ++= (CrossVersion.partialVersion(scalaBinaryVersion.value) match {
        case Some((2, 13)) => Seq.empty[ModuleID]
        case _             => Seq(compilerPlugin(%%("paradise", V.paradise) cross CrossVersion.full))
      })
    )

    lazy val docsDependencies: Def.Setting[Seq[ModuleID]] =
      libraryDependencies += %%("scalatest", V.scalaTest)

    def toCompileTestList(sequence: Seq[ProjectReference]): List[String] = sequence.toList.map {
      p =>
        val project: String = p.asInstanceOf[LocalProject].project
        s"$project/test"
    }
  }

  import autoImport.V

  override def projectSettings: Seq[Def.Setting[_]] =
    Seq(
      name := "github4s",
      orgGithubSetting := GitHubSettings(
        organization = "47degrees",
        project = (name in LocalRootProject).value,
        organizationName = "47 Degrees",
        groupId = "com.47deg",
        organizationHomePage = url("http://47deg.com"),
        organizationEmail = "hello@47deg.com"
      ),
      orgProjectName := "Github4s",
      description := "Github API wrapper written in Scala",
      startYear := Option(2016),
      resolvers += Resolver.sonatypeRepo("snapshots"),
      scalaVersion := V.scala213,
      crossScalaVersions := Seq(V.scala212, V.scala213),
      scalacOptions := {
        val withStripedLinter = scalacOptions.value filterNot Set("-Xlint", "-Xfuture").contains
        (CrossVersion.partialVersion(scalaBinaryVersion.value) match {
          case Some((2, 13)) => withStripedLinter :+ "-Ymacro-annotations"
          case _             => withStripedLinter
        }) :+ "-language:higherKinds"
      },
      orgGithubTokenSetting := "GITHUB_TOKEN",
      orgBadgeListSetting := List(
        TravisBadge.apply(_),
        GitterBadge.apply(_),
        CodecovBadge.apply(_),
        MavenCentralBadge.apply(_),
        LicenseBadge.apply(_),
        ScalaLangBadge.apply(_),
        GitHubIssuesBadge.apply(_)
      ),
      orgScriptTaskListSetting ++= List(
        (ScoverageKeys.coverageAggregate in Test).asRunnableItemFull,
        "docs/mdoc".asRunnableItem
      ),
      coverageExcludedPackages := "<empty>;github4s\\.scalaz\\..*",
      // This is necessary to prevent packaging the BuildInfo with
      // sensible information like the Github token. Do not remove.
      mappings in (Compile, packageBin) ~= { (ms: Seq[(File, String)]) =>
        ms filter {
          case (_, toPath) =>
            !toPath.startsWith("github4s/BuildInfo")
        }
      }
    ) ++ shellPromptSettings ++ sharedScoverageSettings(70d)
}
